# 01 Basic classifier

Here is a starter Jupyter notebook to get started.

## Data exploration

Data exploration is a crucial first step in any data science project.
It involves understanding the data's structure, identifying patterns or trends, and detecting anomalies or missing values.

### Loading data

Loading data is a fundamental step in any data science project, setting the stage for all subsequent analysis and modeling tasks.
Efficiently managing and loading data is especially critical when dealing with large datasets, as improper handling can lead to memory issues and slow performance.
In this section, we will demonstrate how to load our dataset using the [`polars`](https://docs.pola.rs/) and [`pyarrow`](https://arrow.apache.org/docs/python/).

To begin with, we import the [`polars`](https://docs.pola.rs/) library using `import polars as pl`; this high-performance DataFrame library is designed to handle large datasets efficiently.
Polars is known for its speed and memory efficiency, making it an excellent choice for data manipulation and analytical tasks where performance is crucial.
It offers a familiar `DataFrame` API that is intuitive for users with experience in [`pandas`](https://pandas.pydata.org/) but provides significant performance advantages, particularly with large datasets.
Using Polars ensures that our data processing tasks remain performant and scalable, preventing the crashes and slowdowns that often occur with less efficient libraries.

In [1]:
import polars as pl

While `polars` is designed for high-performance data set analysis, there are always limits from the data set size and computational hardware&mdash;specifically memory.
Our Leash biosciences training data is a 3.8 GB [parquet](https://parquet.apache.org/docs/) file which could likely cause issues if trying to load it with `polars`.

Instead, we will demonstrate how to load parquet files with the test set (which is only 30.2 MB).
This is done with the [`read_parquet`](https://docs.pola.rs/py-polars/html/reference/api/polars.read_parquet.html) function in polars.

In [2]:
data_test = pl.read_parquet(source="../../../data/test.parquet")

Once the data is loaded into the DataFrame, we can inspect the first few rows to ensure it has been read correctly and to get an initial understanding of its structure.
This initial inspection helps with:

- **Data Verification**: Ensuring that the data has been loaded correctly is a critical first step. By inspecting the first few rows, we confirm that the data structure matches our expectations.
- **Understanding Structure**: This initial inspection provides a clear picture of the data's structure, which is essential for planning subsequent data processing and analysis steps.
- **Spotting Issues Early**: Early detection of any anomalies or inconsistencies in the data can save significant time and effort later in the analysis process.

The [`DataFrame.head`](https://docs.pola.rs/py-polars/html/reference/dataframe/api/polars.DataFrame.head.html#polars.DataFrame.head) function will print the first five of the DataFrame, giving us a snapshot of the dataset.

In [3]:
print(data_test.head(n=5))

shape: (5, 6)
┌───────────┬─────────────────┬─────────────────┬─────────────────┬─────────────────┬──────────────┐
│ id        ┆ buildingblock1_ ┆ buildingblock2_ ┆ buildingblock3_ ┆ molecule_smiles ┆ protein_name │
│ ---       ┆ smiles          ┆ smiles          ┆ smiles          ┆ ---             ┆ ---          │
│ i64       ┆ ---             ┆ ---             ┆ ---             ┆ str             ┆ str          │
│           ┆ str             ┆ str             ┆ str             ┆                 ┆              │
╞═══════════╪═════════════════╪═════════════════╪═════════════════╪═════════════════╪══════════════╡
│ 295246830 ┆ C#CCCC[C@H](NC( ┆ C=Cc1ccc(N)cc1  ┆ C=Cc1ccc(N)cc1  ┆ C#CCCC[C@H](Nc1 ┆ BRD4         │
│           ┆ =O)OCC1c2ccccc2 ┆                 ┆                 ┆ nc(Nc2ccc(C=C)c ┆              │
│           ┆ …               ┆                 ┆                 ┆ …               ┆              │
│ 295246831 ┆ C#CCCC[C@H](NC( ┆ C=Cc1ccc(N)cc1  ┆ C=Cc1ccc(N)cc1  ┆ C#CCCC[C@

After loading the test data into a Polars DataFrame, we inspect the first few rows to verify the data's integrity and understand its structure.
The output reveals that the DataFrame has a shape of 5 rows and 6 columns, which aligns with our request to display the first five rows.
This initial check confirms that the DataFrame has the expected dimensions, which is a critical step in validating that the data has been loaded correctly.

Each column in the DataFrame serves a specific purpose and has a designated data type.
The `id` column is a unique identifier for each molecule-protein pair, and it is stored as a 64-bit integer (`i64`).
The `buildingblock1_smiles`, `buildingblock2_smiles`, and `buildingblock3_smiles` columns contain the SMILES representations of the individual building blocks of the molecules.
These columns are of type `str` (string), as SMILES strings are textual representations of chemical structures.
The `molecule_smiles` column holds the SMILES representation of the fully assembled molecule, which includes the three building blocks and the core structure.
This column is also of type `str`. Lastly, the `protein_name` column, which is of type `str`, indicates the name of the protein target that each molecule is associated with.

The content of the rows provides a snapshot of the data, where each row represents a unique molecule-protein pair along with its corresponding SMILES strings and protein name.
For instance, the first row shows a molecule with an `id` of 295246830, which is composed of three building blocks with their respective SMILES strings.
The fully assembled molecule's SMILES string is also provided, and the protein target for this molecule is `BRD4`. This detailed breakdown allows us to understand the composition and target of each molecule in the dataset.

Additionally, we notice the presence of ellipses (`…`) in the SMILES strings.
These ellipses indicate that the SMILES strings are too long to be fully displayed in the output, which is common given the complexity and length of chemical structures represented in SMILES notation.
The truncated display helps maintain readability while providing a glimpse of the data's content.

We can also look at the last two rows using the [`DataFrame.tail`](https://docs.pola.rs/py-polars/html/reference/dataframe/api/polars.DataFrame.tail.html) function.
Now we have to specify `n=2` to change the number of rows.

In [4]:
print(data_test.tail(n=2))

shape: (2, 6)
┌───────────┬─────────────────┬─────────────────┬─────────────────┬─────────────────┬──────────────┐
│ id        ┆ buildingblock1_ ┆ buildingblock2_ ┆ buildingblock3_ ┆ molecule_smiles ┆ protein_name │
│ ---       ┆ smiles          ┆ smiles          ┆ smiles          ┆ ---             ┆ ---          │
│ i64       ┆ ---             ┆ ---             ┆ ---             ┆ str             ┆ str          │
│           ┆ str             ┆ str             ┆ str             ┆                 ┆              │
╞═══════════╪═════════════════╪═════════════════╪═════════════════╪═════════════════╪══════════════╡
│ 296921724 ┆ [N-]=[N+]=NCCC[ ┆ Nc1noc2ccc(F)cc ┆ NCc1cccs1       ┆ [N-]=[N+]=NCCC[ ┆ HSA          │
│           ┆ C@H](NC(=O)OCC1 ┆ 12              ┆                 ┆ C@H](Nc1nc(NCc2 ┆              │
│           ┆ …               ┆                 ┆                 ┆ …               ┆              │
│ 296921725 ┆ [N-]=[N+]=NCCC[ ┆ Nc1noc2ccc(F)cc ┆ NCc1cccs1       ┆ [N-]=[N+]

### Chunked reading

For very large datasets, chunked reading is a more efficient approach.
This method reads the dataset in smaller chunks, which helps manage memory usage more effectively.
By processing data incrementally, chunked reading avoids the pitfalls of memory overload and allows for scalable data exploration and analysis.

1. **Memory Management**: Large datasets can easily exceed the available memory, causing kernel crashes or significant slowdowns. Chunked reading mitigates this by only loading manageable portions of the data at a time.
2. **Scalability**: Chunked reading enables the handling of datasets that are much larger than the system’s memory capacity, making it a scalable solution for big data problems.
3. **Incremental Processing**: It allows for incremental processing of data, which can be useful for real-time data analysis and processing tasks.
4. **Flexibility**: You can perform operations on each chunk independently, which provides flexibility in data processing and can lead to more efficient computations.

Chunked reading involves reading a fixed number of rows (a chunk) from the dataset, processing that chunk, and then moving on to the next chunk.
This approach ensures that only a small portion of the dataset is loaded into memory at any given time.
To start, we need to import the necessary library for reading parquet files.
We will use the [`pyarrow.parquet`](https://arrow.apache.org/docs/python/index.html) library to handle the parquet files.

In [5]:
import pyarrow.parquet as pq

With the library imported, we can now define a function to read the dataset in chunks.
This will help us manage the large data size without overwhelming our system's memory.


[ParqetFile](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetFile.html)

In [6]:
path_train_data = "../../../data/train.parquet"
data_train = pq.ParquetFile(source=path_train_data)

For very large datasets, using `pyarrow.ParquetFile` with the `iter_batches` method is an efficient approach to manage memory usage.
This method reads the dataset in smaller chunks or batches, allowing for scalable data exploration and analysis.
We will define a function to read the dataset in batches using `pyarrow`.
This function will handle reading the data in specified batch sizes.

In [7]:
data_train_gen = data_train.iter_batches()

To ensure our batch reading function works correctly, we will retrieve and inspect the first batch of data.

In [8]:
data_train_sample = next(data_train_gen)
print(data_train_sample)

pyarrow.RecordBatch
id: int64
buildingblock1_smiles: string
buildingblock2_smiles: string
buildingblock3_smiles: string
molecule_smiles: string
protein_name: string
binds: int64
----
id: [0,1,2,3,4,5,6,7,8,9,...,65526,65527,65528,65529,65530,65531,65532,65533,65534,65535]
buildingblock1_smiles: ["C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21","C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21","C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21","C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21","C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21","C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21","C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21","C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21","C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21","C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21",...,"C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21","C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21","C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21","C#CC[C@@H](CC(=O)O)NC(=O)OCC1c2ccccc2-c2ccccc21"